vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/LinkManager.java

```diff
@@ -89,182 +36,72 @@
+    public ILinkDisplay updateLink(String title, String id, float positionX, float positionY,
+            int rotation, String linkedId, String linkLabel, String linkId, String linkDisplayId,
+            DisplayType displayType, byte[] linkImage, String imageFilename) throws SpaceDoesNotExistException, LinkDoesNotExistsException, ImageCouldNotBeStoredException{
 
-        if (linkImage != null && linkImage.length > 0) {
-            Tika tika = new Tika();
-            String contentType = tika.detect(linkImage);
-
-            IVSImage image = imageFactory.createImage(imageFilename, contentType);
-            image = imageRepo.save((VSImage) image);
+        spaceValidation(id);
 
-            String relativePath = null;
-            try {
-                relativePath = storage.storeFile(linkImage, imageFilename, image.getId());
-            } catch (FileStorageException e) {
-                throw new ImageCouldNotBeStoredException(e);
-            }
-            image.setParentPath(relativePath);
-            imageRepo.save((VSImage) image);
+        ILink link = getLink(linkId);
+        ILinkDisplay displayLink = getDisplayLink(linkDisplayId);
 
-            display.setImage(image);
-        }
 
-        spaceLinkDisplayRepo.save((SpaceLinkDisplay) display);
-        return display;
-    }
+        IVSpaceElement target = getTarget(linkedId);
+        link.setName(title);
+        setTarget(link,target);
 
-    @Override
-    public void deleteSpaceLink(String linkId) {
-        Optional<SpaceLink> linkOptional = spaceLinkRepo.findById(linkId);
-        if (!linkOptional.isPresent()) {
-            return;
-        }
+        setDisplayProperties(displayLink,positionX,positionY,rotation, displayType, linkImage, imageFilename);
 
-        ISpace space = linkOptional.get().getSourceSpace();
-        ISpaceLink link = linkOptional.get();
-        space.getSpaceLinks().remove(link);
-        spaceLinkDisplayRepo.deleteByLink(link);
-        spaceLinkRepo.delete((SpaceLink) link);
+        return updateLinkAndDisplay(link,displayLink);
     }
 
-    @Override
-    public void deleteModuleLink(String linkId) {
-        Optional<ModuleLink> linkOptional = moduleLinkRepo.findById(linkId);
-        if (!linkOptional.isPresent()) {
-            return;
-        }
+    protected abstract ILinkDisplay updateLinkAndDisplay(ILink link, ILinkDisplay displayLink);
 
-        ISpace space = linkOptional.get().getSpace();
-        IModuleLink link = linkOptional.get();
-        space.getModuleLinks().remove(link);
-        moduleLinkDisplayRepo.deleteByLink(link);
-        moduleLinkRepo.delete((ModuleLink) link);
-    }
-    
-    @Override
-    public void deleteExternalLink(String linkId) {
-        Optional<ExternalLink> linkOptional = externalLinkRepo.findById(linkId);
-        if (!linkOptional.isPresent()) {
-            return;
-        }
+    protected abstract void setTarget(ILink link, IVSpaceElement target);
```

Comment:
thinking about this, i would actually use generics in this class as well. this way this method could be
```
setTarget(L link, T target)
```
where L extends ILink and T extends IVSpaceElement. You could avoid a lot of casting in the subclasses.

Reply:
everywhere you use IVSpaceElement, it should use T